File: GeoBroker-Commons/src/main/java/de/hasenburg/geobroker/commons/model/spatial/Location.java
package de.hasenburg.geobroker.commons.model.spatial;

import com.fasterxml.jackson.annotation.JsonCreator;
import com.fasterxml.jackson.annotation.JsonIgnore;
import com.fasterxml.jackson.annotation.JsonProperty;
import de.hasenburg.geobroker.commons.Utility;
import de.hasenburg.geobroker.commons.model.JSONable;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import org.jetbrains.annotations.NotNull;
import org.locationtech.spatial4j.distance.DistanceUtils;
import org.locationtech.spatial4j.io.ShapeWriter;
import org.locationtech.spatial4j.shape.Point;

import java.text.ParseException;
import java.util.Objects;
import java.util.Random;

import static de.hasenburg.geobroker.commons.model.spatial.SpatialContext.GEO;
import static org.locationtech.spatial4j.distance.DistanceUtils.DEG_TO_KM;
import static org.locationtech.spatial4j.distance.DistanceUtils.KM_TO_DEG;

public class Location implements JSONable {

	private static final Logger logger = LogManager.getLogger();

	private final Point point;

	private boolean undefined = false;

	private Location(Point point) {
		this.point = point;
	}

	private Location(boolean undefined) {
		this.undefined = undefined;
		this.point = null;
	}

	@JsonCreator
	private Location(@JsonProperty("WKT") String wkt) throws ParseException {
		if ("{ undefined }".equals(wkt)) {
			this.point = null;
			this.undefined = true;
		} else {
			WKTReader reader = (WKTReader) GEO.getFormats().getWktReader();
			this.point = (Point) reader.parse(wkt);
		}
	}

	/**
	 * Creates a location with the given lat/lon coordinates.
	 *
	 * @param lat - the latitude (Breitengrad)
	 * @param lon - the longitude (Längengrad)
	 */
	public Location(double lat, double lon) {
		point = GEO.getShapeFactory().pointLatLon(lat, lon);
	}

	/**
	 * Creates an undefined location, i.e., one that does not have a point and is not contained in any geofence.
	 */
	@NotNull
	public static Location undefined() {
		return new Location(true);
	}

	/**
	 * Creates a random location (Not inclusive of (-90, 0))
	 */
	public static Location random() {
		Random random = new Random();
		// there have been rounding errors
		return new Location(Math.min((random.nextDouble() * -180.0) + 90.0, 90.0),
				Math.min((random.nextDouble() * -360.0) + 180.0, 180.0));
	}

	/**
	 * Creates a random location that is inside the given Geofence.
	 *
	 * @param geofence - may not be a geofence that crosses any datelines!!
	 * @return a random location or null if the geofence crosses a dateline
	 */
	public static Location randomInGeofence(Geofence geofence) {
		Location result = null;
		int i = 0;
		do {
			// generate lat in bounding box
			double lat = Utility.randomDouble(geofence.getBoundingBoxSouthWest().getLat(),
					geofence.getBoundingBoxNorthEast().getLat());

			// generate lon in bounding box
			double lon = Utility.randomDouble(geofence.getBoundingBoxSouthWest().getLon(),
					geofence.getBoundingBoxNorthEast().getLon());

			// create location and hope it is in geofence
			result = new Location(lat, lon);
		} while (!geofence.contains(result) && ++i < 1000);
		// location was in geofence, so let's return it
		return result;
	}

	/**
	 * @param location - starting location
	 * @param distance - distance from starting location in km
	 * @param direction - direction (0 - 360)
	 */
	public static Location locationInDistance(Location location, double distance, double direction) {
		if (location.isUndefined()) {
			return Location.undefined();
		}
		Point result = GEO.getDistCalc().pointOnBearing(location.point,
				distance * KM_TO_DEG,
				direction,
				GEO,
				GEO.getShapeFactory().pointLatLon(0.0, 0.0));

		return new Location(result);
	}

	/**
	 * Distance between this location and the given one, as determined by the Haversine formula, in radians
	 *
	 * @param toL - the other location
	 * @return distance in radians or -1 if one location is undefined
	 */
	public double distanceRadiansTo(Location toL) {
		if (undefined || toL.undefined) {
			return -1.0;
		}
		return GEO.getDistCalc().distance(point, toL.getPoint());
	}

	/**
	 * Distance between this location and the given one, as determined by the Haversine formula, in km
	 *
	 * @param toL - the other location
	 * @return distance in km or -1 if one location is undefined
	 */
	public double distanceKmTo(Location toL) {
		if (undefined || toL.undefined) {
			return -1.0;
		}
		return distanceRadiansTo(toL) * DEG_TO_KM;
	}

	/*****************************************************************
	 * Getters and String
	 ****************************************************************/

	@JsonIgnore
	public Point getPoint() {
		return point;
	}

	@JsonIgnore
	public double getLat() {
		return point.getLat();
	}

	@JsonIgnore
	public double getLon() {
		return point.getLon();
	}

	@JsonProperty("WKT")
	public String getWKTString() {
		if (undefined) {
			return "{ undefined }";
		}
		ShapeWriter writer = GEO.getFormats().getWktWriter();
		return writer.toString(point);
	}

	@JsonIgnore
	public boolean isUndefined() {
		return undefined;
	}

	@Override
	public String toString() {
		return getWKTString();
	}

	/*****************************************************************
	 * Generated methods
	 ****************************************************************/

	@Override
	public boolean equals(Object o) {
		if (this == o) return true;
		if (o == null || getClass() != o.getClass()) return false;
		Location location = (Location) o;
		return undefined == location.undefined && Objects.equals(point, location.point);
	}

	@Override
	public int hashCode() {
		return Objects.hash(point, undefined);
	}

	public static void main(String[] args) {
		Location l = new Location(39.984702, 116.318417);
		Location l2 = new Location(39.974702, 116.318417);
		logger.info("Distance is {}km", l.distanceKmTo(l2));

		l = new Location(57.34922076607738, 34.53035122251791);
		l2 = new Location(57.34934475583778, 34.53059311887825);
		logger.info("Distance is {}km", l.distanceKmTo(l2));
	}
}

File: GeoBroker-Commons/src/main/kotlin/de/hasenburg/geobroker/commons/model/spatial/Location.kt
package de.hasenburg.geobroker.commons.model.spatial

import kotlinx.serialization.*
import kotlinx.serialization.internal.StringDescriptor
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonConfiguration
import org.apache.logging.log4j.LogManager
import org.locationtech.spatial4j.distance.DistanceUtils
import org.locationtech.spatial4j.shape.Point
import kotlin.math.min
import kotlin.random.Random

private val logger = LogManager.getLogger()

@Serializable
class Location(@Serializable(with = PointWKTSerializer::class) @SerialName("wkt") val point: Point) {

    /**
     * Creates a location with the given lat/lon coordinates.
     *
     * @param lat - the latitude (Breitengrad)
     * @param lon - the longitude (Längengrad)
     */
    constructor(lat: Double, lon: Double) : this(SpatialContextK.GEO.shapeFactory.pointLatLon(lat, lon))

    val lat: Double
        get() = point.lat

    val lon: Double
        get() = point.lon

    /**
     * @param distance - distance from starting location in km
     * @param direction - direction (0 - 360)
     */
    fun locationInDistance(distance: Double, direction: Double): Location {
        return Location(SpatialContextK.GEO.distCalc.pointOnBearing(this.point,
                distance * DistanceUtils.KM_TO_DEG,
                direction,
                SpatialContextK.GEO,
                SpatialContextK.GEO.shapeFactory.pointLatLon(0.0, 0.0)))
    }

    /**
     * Distance between this location and the given one, as determined by the Haversine formula, in radians
     *
     * @param toL - the other location
     * @return distance in radians
     */
    fun distanceRadiansTo(toL: Location): Double {
        return SpatialContextK.GEO.distCalc.distance(point, toL.point)
    }

    /**
     * Distance between this location and the given one, as determined by the Haversine formula, in km
     *
     * @param toL - the other location
     * @return distance in km or -1 if one location is undefined
     */
    fun distanceKmTo(toL: Location): Double {
        return distanceRadiansTo(toL) * DistanceUtils.DEG_TO_KM
    }

    companion object {
        /**
         * Creates a random location (Not inclusive of (-90, 0))
         */
        @JvmOverloads
        fun random(random: Random = Random.Default): Location {
            // there have been rounding errors
            return Location(min((random.nextDouble() * -180.0) + 90.0, 90.0),
                    min((random.nextDouble() * -360.0) + 180.0, 180.0))
        }

        /**
         * Creates a random location that is inside the given Geofence.
         *
         * @param geofence - may not be a geofence that crosses any datelines!!
         * @return a random location or null if the geofence crosses a dateline
         */
        @JvmOverloads
        fun randomInGeofence(geofence: Geofence, random: Random = Random.Default): Location? {
            var result: Location
            var i = 0
            do {
                // generate lat/lon in bounding box
                val lat = random.nextDouble(geofence.boundingBoxSouthWest.lat, geofence.boundingBoxNorthEast.lat)
                val lon = random.nextDouble(geofence.boundingBoxSouthWest.lon, geofence.boundingBoxNorthEast.lon)

                // create location and hope it is in geofence
                result = Location(lat, lon)
            } while (!geofence.contains(result) && ++i < 1000)

            return if (geofence.contains(result)) {
                // location was in geofence, so let's return it
                result
            } else {
                // none was found
                null
            }
        }

    }

    /*****************************************************************
     * Generated methods
     ****************************************************************/

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false

        other as Location

        if (point.lat != other.point.lat) return false

        return true
    }

    override fun hashCode(): Int {
        return point.hashCode()
    }

    override fun toString(): String {
        return "Location(lat=$lat, lon=$lon)"
    }

}

/*****************************************************************
 * Json Serialization
 ****************************************************************/

fun Location.toJson(json: Json = Json(JsonConfiguration.Stable)): String {
    return json.stringify(Location.serializer(), this)
}

/**
 * @throws [kotlinx.serialization.json.JsonDecodingException]
 */
fun String.toLocation(json: Json = Json(JsonConfiguration.Stable)): Location {
    return json.parse(Location.serializer(), this)
}

object PointWKTSerializer : KSerializer<Point> {

    override val descriptor: SerialDescriptor = StringDescriptor

    override fun serialize(encoder: Encoder, obj: Point) {
        encoder.encodeString(SpatialContextK.GEO.formats.wktWriter.toString(obj))
    }

    override fun deserialize(decoder: Decoder): Point {
        val reader = SpatialContextK.GEO.formats.wktReader as CustomWKTReader
        return reader.parse(decoder.decodeString()) as Point
    }
}
